### PR TITLE
A fun tool for painting things

### PR DESCRIPTION
**Pull request by seijihariki**
A tool to use your master Microsoft Paint® artist skills to help you on your Stable Diffusion endeavors. 

Not much to say about it, really. Other than, the color picker of browsers doesn't seem to pick colors correctly with the stacked canvas system we are using, so we'll have to make our own.

Basically, the mask brush tool but for the image layer. When painting, it also (optionally) paints the equivalent mask (though I did not check if that works well with the blur):

https://user-images.githubusercontent.com/7552751/205165742-1b6fa777-4783-400a-ad76-fc4516e0a42d.mp4

Also works to give some better initial instructions

https://user-images.githubusercontent.com/7552751/205169129-72f2e792-8533-4e98-9a81-9ee5084ff636.mp4

Someone with a better drawing hand than I would probably do a much better job at this.

### CODE REVIEW

**Comment by zero01101** [comment]
i could kiss you

**Comment by zero01101** [comment]
>(though I did not check if that works well with the blur):

lol the auto-color-brush mask definitely gets eaten by overmask, terribly sorry about that :badpokerface:

**Comment by seijihariki** [comment]
These last two commits add just actions (rename/delete) to resources, and updates the connection status indicator a bit

**Comment by zero01101** [comment]
hah, my last commit was to just ensure that `stored` isn't null if the json.parse has nothing in localstorage to parse :D  loving all of this so far

**Review by zero01101 — APPROVED** [review]
minus the unsolved mask brush being left behind if a color brush is "undone" instead of erased, this is all magnificent and too much fun

**Comment by seijihariki** [comment]
> minus the unsolved mask brush being left behind if a color brush is "undone" instead of erased, this is all magnificent and too much fun

That was kind of deliberate, as we kinda stablished that the history does nothing to the mask with our 

> 
> [edit] is there any way to _unselect_ a style? might need a "none" option added back in

That's a good point! I did not have much time to test those features, as you can see by my very professional and clear commit titles.

It seems you have developed a fix! I'll give that a look as soon as I can.

**Comment by zero01101** [comment]
totally agreed re: "idempotency" of masks and _really_ solid point about mixed mask layer history 😬 i just foresee a potential user (read as: me) undoing a sketch and getting the mask results unexpectedly down the line lol, but that sounds like a job for docs more than anything ;) 

> as you can see by my very professional and clear commit titles.

lmao have you seen some of the garbage i emit

File: js/index.js [hunk 1]
```diff
@@ -803,14 +808,21 @@ function changeStyles() {
...
+		selectedString = [];
+		Array.from(styleSelectEl.options).forEach((option) => {
+			if (option.value !== "None") option.selected = false;
+		});
+	}
 
-	selectedString != "None"
-		? localStorage.setItem("promptStyle", JSON.stringify(selectedString))
-		: localStorage.setItem("promptStyle", "[]");
+	localStorage.setItem("promptStyle", JSON.stringify(selectedString));
```
Comment (seijihariki) [inline js/index.js:820]: Yeah, my naming conventions were not really clear... selectedString is actually an array of strings. This comparison may have worked due to automagic JavaScript casting, but I wouldn't bet on the consistency of that.

Comment (seijihariki) [reply js/index.js:820]: This last commit now fixes this, but also clears all selections if "None" is ever selected, as well as adding a helpful check indicator on applied styles:
image

None in action:

https://user-images.githubusercontent.com/7552751/205426710-71322f8e-b3f2-49b9-a96b-1522bf56869c.mp4

**Comment by seijihariki** [comment]
This last one is quite interesting. I felt the mouse wheel was lonely on the dream() tools. So, I did something with it.

Adds dream reticle size separated from Stable Diffusion resolution. That makes it possible to manually do what full resolution inpainting does, and makes us able to better add smaller details to a scene. Also makes it possible to replicate auto TXT to image fix functionality.

Specific region inpainting:

https://user-images.githubusercontent.com/7552751/205442690-b90982ad-1366-49ef-8264-1937f9d28df0.mp4

Big image generation, then local upscaling:

https://user-images.githubusercontent.com/7552751/205442830-efa9cb46-9f74-488b-a623-8d7e50c03c16.mp4

But basically, this functionality gives us much more freedom to generate images and insert them into the world.

Obs: Please don't mind that it always generates women. For some reason, the Beautiful # styles I have always end up doing women unless I specify.

**Comment by zero01101** [comment]
decoupling the requested generation resolution from the resultant output resolution is super clever, my only complaint is that i'll have to rewrite some readme fundamentals ;)

**Review by zero01101 — CHANGES_REQUESTED** [review]
all systems go minus dream.js:432 and :434 :)

File: js/ui/tool/dream.js [hunk 2]
```diff
@@ -374,23 +393,45 @@ const dream_generate_callback = async (evn, state) => {
...
 				// 99% of my issues were from failing to set source-over for the overmask blotches
 				if (state.overMaskPx > 0) {
 					// transparent to white first
 					auxCtx.globalCompositeOperation = "destination-atop";
 					auxCtx.fillStyle = "#FFFF";
-					auxCtx.fillRect(0, 0, bb.w, bb.h);
+					auxCtx.fillRect(0, 0, request.width, request.height);
+					downloadCanvas({canvas: auxCanvas, filename: null});
 					applyOvermask(auxCanvas, auxCtx, state.overMaskPx);
+					downloadCanvas({canvas: auxCanvas, filename: null});
```
[the remark this answers is not in the document]
Comment (seijihariki) [reply js/ui/tool/dream.js:434]: That's a good point!

**Comment by seijihariki** [comment]
Will merge here then

**Comment by zero01101** [comment]
omg you even added error handling for failed dreams you are a hero

**Pull request merged**